File: sysplant/templates/resolvers/random.nim
proc SPT_GetRandomSyscallAddress(hash: DWORD): ULONG_PTR =
    let entries = toSeq(ssdt.values)
    # Pickup random non-hooked syscall to JUMP to
    var entry: Syscall = entries.sample()
    
    # Avoid random bad luck...
    while entry.ssn == ssdt[hash].ssn:
        entry = entries.sample()

    return cast[ULONG_PTR](entry.address)

File: sysplant/templates/resolvers/random.c
EXTERN_C PVOID SPT_GetRandomSyscallAddress(DWORD FunctionHash)
{
    // Ensure SPT_SyscallList is populated.
    if (!SPT_PopulateSyscallList()) return NULL;

    DWORD index;

    while (TRUE){
        // Pickup random non-hooked syscall to JUMP to
        index = ((DWORD) rand()) % SPT_SyscallList.Count;
        // Avoid random bad luck...
        if (FunctionHash == SPT_SyscallList.Entries[index].Hash) {
            continue;
        }
        // Do not choose hooked functions
        if (SPT_SyscallList.Entries[index].SyscallAddress == SPT_SyscallList.Entries[index].Address) {
            continue;
        }
        // Otherwise it's all good
        break;
    }
    
    return SPT_SyscallList.Entries[index].SyscallAddress;
}
